### PR TITLE
[RLlib] Fix PPOTorchPolicy producing float metrics when not using critic

### LINKED ISSUE #27822: [RL Lib] Reporting metric / stats wrong type.
### What happened + What you expected to happen

We found a bug in PPOTorchPolicy when using
use_critic = False
Bug is on lines 173 and 174.
        model.tower_stats["mean_vf_loss"] = mean_vf_loss
        model.tower_stats["vf_explained_var"] = explained_variance(
            train_batch[Postprocessing.VALUE_TARGETS], value_fn_out
        )
On 173 wrong type is used for reporting (Float instead of tensor).
On 174 the the explained_variance probably should not be called without the critic.
Hotfix that works for me:
       vf_explained_var = explained_variance(
            train_batch[Postprocessing.VALUE_TARGETS], value_fn_out
        ) if self.config["use_critic"] else torch.Tensor([0.0])
        mean_vf_loss_stats = mean_vf_loss if self.config["use_critic"] else torch.Tensor([0.0])

### Versions / Dependencies

1.13

### Reproduction script

see above. if you need additional context from customer, reach out to me.

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Signed-off-by: Artur Niederfahrenhorst <artur@anyscale.com>

## Why are these changes needed?

PPOTorchPolicy produces float metrics when not using a critic. This produces errors when, for example, moving them to another device. This also unifies the behaviour with TF, which will default to vf_explained_var=-1.

## Related issue number

Closes #27822 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/torch_utils.py [hunk 1]
```diff
@@ -211,6 +211,9 @@ def explained_variance(y: TensorType, pred: TensorType) -> TensorType:
         The explained variance given a pair of labels and predictions.
     """
     y_var = torch.var(y, dim=[0])
+    if y_var == 0.0:
+        # Model case in which y does not vary with explained variance of -1
+        return torch.tensor(-1.0)
```
Comment (sven1977) [inline rllib/utils/torch_utils.py:216]: We should do `torch.tensor(-1.0).to(pred.device)` instead to make sure this new code is also GPU compatible.

File: rllib/algorithms/ppo/ppo_torch_policy.py [hunk 2]
```diff
@@ -148,8 +148,8 @@ def reduce_mean_valid(t):
             mean_vf_loss = reduce_mean_valid(vf_loss_clipped)
         # Ignore the value function.
         else:
-            value_fn_out = 0
-            vf_loss_clipped = mean_vf_loss = 0.0
+            value_fn_out = torch.tensor(0.0).to(self.device)
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ppo/ppo_torch_policy.py]: Done. Looking at TorchPolicyV2 though, I think self.device will only be "cpu" in the case of fake GPUs or no GPUs.

`self.devices = [
                torch.device("cuda:{}".format(i))
                for i, id_ in enumerate(gpu_ids)
                if i < num_gpus
            ]
 self.device = self.devices[0]`

**Review by sven1977 — APPROVED** [review]
Thanks for the fixes @ArturNiederfahrenhorst , looks good now.

File: rllib/algorithms/ppo/ppo_torch_policy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/ppo/ppo_torch_policy.py]: But still, if you had multiple GPUs, which one would be stored in `self.device`?
It's better to definitely use the same device as all the other tensors in this particular instance of the loss function (which is called on each(!) of the GPUs).
